src/crimson/net/ProtocolV2.cc

```diff
@@ -179,6 +181,38 @@ seastar::future<> ProtocolV2::write_flush(bufferlist&& buf)
   return socket->write_flush(std::move(buf));
 }
 
+seastar::temporary_buffer<char>
+ProtocolV2::create(seastar::compat::polymorphic_allocator<char>* const allocator)
+{
+#if 0
+  // space for segments_n, epilogue_n and preable_n+1
+  return seastar::make_temporary_buffer<char>(allocator, 8192);
+#else
+
+  if (!last_returned.empty()) {
+    return std::move(last_returned);
+  }
+
+  if (rx_segments_desc.empty()) {
+    // space just for the ceph banner + preamble_0
+    // FIXME: magic
+    return seastar::make_temporary_buffer<char>(allocator, 26 + FRAME_PREAMBLE_SIZE);
+  } else {
+    // space for segments_n, epilogue_n and preable_n+1
+    size_t segment_size_sum = 0;
+    for (const auto& segment : rx_segments_desc) {
+      segment_size_sum += segment.length;
+    }
+    return seastar::make_temporary_buffer<char>(allocator,
+        segment_size_sum + FRAME_PLAIN_EPILOGUE_SIZE + FRAME_PREAMBLE_SIZE);
+  }
+
+  // TODO: implement prefetching for very small (under 4K) chunk sizes to not
```

Comment:
yeah, there is a gray area we could benefit from trading less memcpy() for less read() syscalls using prefetching. but the question is, how can we predict the sizes of following segments? do we have a "hint" for the allocator in the v2 protocol?

Reply:
> do we have a "hint" for the allocator in the v2 protocol?

I don't think so.

> the question is, how can we predict the sizes of following segments?

The simplest idea would be to extrapolate – if the prefetch buffer needs refill and the current msg size is small, allocate a bigger chunk. In the best scenario next messages are also small. If not, then we can handle the upcoming bigger one with SGL (let it be fragmented and even unaligned) to defer the rectifying `memcpy` to e.g. kernel handling `io_submit()` (or compression/encryption/* shuffling the memory).

> yeah, there is a gray area where we could benefit from trading less memcpy() for less read() syscalls using prefetching.

Prefetching can be quirky. I'm afraid of the next problem: the threshold (and its evolution over time). It depends on many things. DPDK vs POSIX, hardware (security bugs, shuffling performance), kernel version (workarounds for them making mode-switch costlier; `io_uring` support), ...